# it is a regression problem 

## we are doing the  ml on house pricing 

<table align="center" color="#216180" width=100%>
    <tr>
        </td>
            </td>
        </td>
        <td>
            <div align="center">
                <font color="#216180" size=16px>
                    <b> Regression modeling on House Price Data
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

# Problem Statement

This dataset is created for the prediction of House Prices from an Indian perspective Towns. The dataset contains various features that are important during the application for the selling the Houses. The predicted output obtained from the classification algorithm gives a fair idea about the chances of a Houses Sold.

In [1]:
#import numpy
import numpy as np#to find statestical summery 
#import pandas
import pandas as pd#formate the data stacture (convert any data stucture into pandas data frame)
#import matplotlib
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap
#import seaborn
import seaborn as sn

### load the data set

### load the data and name it df_house_price

In [2]:
#load csv file
#stire the data in df_houseprice
df_houseprice = pd.read_csv(r"C:\Users\Prachi Prakhar\Downloads\House_Price1.csv")
df_houseprice

price  resid_area  air_qual  room_num   age  distance  teachers  \
0     24.0       32.31     0.538     6.575  65.2      4.35      24.7   
1     21.6       37.07     0.469     6.421  78.9      4.99      22.2   
2     34.7       37.07     0.469     7.185  61.1      5.03      22.2   
3     33.4       32.18     0.458     6.998  45.8      6.21      21.3   
4     36.2       32.18     0.458     7.147  54.2      6.16      21.3   
..     ...         ...       ...       ...   ...       ...       ...   
501   22.4       41.93     0.573     6.593  69.1      2.64      19.0   
502   20.6       41.93     0.573     6.120  76.7      2.44      19.0   
503   23.9       41.93     0.573     6.976  91.0      2.34      19.0   
504   22.0       41.93     0.573     6.794  89.3      2.54      19.0   
505   19.0       41.93     0.573     6.030  80.8      2.72      19.0   

     poor_prop airport  n_hos_beds  n_hot_rooms       waterbody  rainfall  \
0         4.98     YES       5.480      11.1920           River        23   
1         9.14      NO       7.332      12.1728            Lake        42   
2         4.03      NO       7.394     101.1200             NaN        38   
3         2.94     YES       9.268      11.2672            Lake        45   
4         5.33      NO       8.824      11.2896            Lake        55   
..         ...     ...         ...          ...             ...       ...   
501       9.67      NO       9.348      12.1792  Lake and River        27   
502       9.08     YES       6.612      13.1648  Lake and River        20   
503       5.64      NO       5.478      12.1912             NaN        31   
504       6.48     YES       7.940      15.1760             NaN        47   
505       7.88     YES      10.280      10.1520             NaN        45   

    bus_ter     parks  Sold  
0       YES  0.049347     0  
1       YES  0.046146     1  
2       YES  0.045764     0  
3       YES  0.047151     0  
4       YES  0.039474     0  
..      ...       ...   ...  
501     YES  0.056006     1  
502     YES  0.059903     1  
503     YES  0.057572     1  
504     YES  0.060694     1  
505     YES  0.060336     1  

[506 rows x 16 columns]

In [3]:
#name of columns
df_houseprice.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

**interpretation**

 - `price` :price of the house.
 - `resid_area` :residentially area where my house is located

In [4]:
# basic info of the data set
df_houseprice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   price        506 non-null    float64
 1   resid_area   506 non-null    float64
 2   air_qual     506 non-null    float64
 3   room_num     506 non-null    float64
 4   age          506 non-null    float64
 5   distance     506 non-null    float64
 6   teachers     506 non-null    float64
 7   poor_prop    506 non-null    float64
 8   airport      506 non-null    object 
 9   n_hos_beds   498 non-null    float64
 10  n_hot_rooms  506 non-null    float64
 11  waterbody    351 non-null    object 
 12  rainfall     506 non-null    int64  
 13  bus_ter      506 non-null    object 
 14  parks        506 non-null    float64
 15  Sold         506 non-null    int64  
dtypes: float64(11), int64(2), object(3)
memory usage: 63.4+ KB


**interpretation**
 - we have 13 numerical columns
 - we have NULL value in data set
 - we have 16 column in data

In [5]:
#shape of the data
df_houseprice.shape

(506, 16)

**interpretation**
 - we have total `16` columns
 - we have total `506` objervations

In [6]:
# summaraze the badic stastic of data
df_houseprice.describe()

price  resid_area    air_qual    room_num         age    distance  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    22.528854   41.136779    0.554695    6.284634   68.574901    3.971996   
std      9.182176    6.860353    0.115878    0.702617   28.148861    2.108532   
min      5.000000   30.460000    0.385000    3.561000    2.900000    1.130000   
25%     17.025000   35.190000    0.449000    5.885500   45.025000    2.270000   
50%     21.200000   39.690000    0.538000    6.208500   77.500000    3.385000   
75%     25.000000   48.100000    0.624000    6.623500   94.075000    5.367500   
max     50.000000   57.740000    0.871000    8.780000  100.000000   12.320000   

         teachers   poor_prop  n_hos_beds  n_hot_rooms    rainfall  \
count  506.000000  506.000000  498.000000   506.000000  506.000000   
mean    21.544466   12.653063    7.899767    13.041605   39.181818   
std      2.164946    7.141062    1.476683     5.238957   12.513697   
min     18.000000    1.730000    5.268000    10.057600    3.000000   
25%     19.800000    6.950000    6.634500    11.189800   28.000000   
50%     20.950000   11.360000    7.999000    12.720000   39.000000   
75%     22.600000   16.955000    9.088000    14.170800   50.000000   
max     27.400000   37.970000   10.876000   101.120000   60.000000   

            parks        Sold  
count  506.000000  506.000000  
mean     0.054454    0.454545  
std      0.010632    0.498422  
min      0.033292    0.000000  
25%      0.046464    0.000000  
50%      0.053507    0.000000  
75%      0.061397    1.000000  
max      0.086711    1.000000

**interpretation**
 - we have a outliers in the data most of the outlier aer in age column

## data preprocessing


## a. Data Cleaning

In [7]:
#show me 1st FIVE observation
df_houseprice.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200       NaN        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

In [8]:
#step2 alllocate new colum name's to old colums name
#make the Dictonary{"old column name":"new column name"}
D1 ={'price' : 'Price',
     'resid_area' : 'Residential_Area',
     'air_qual': 'Air_Qulity',
    'room_num': 'Room_Numbers',
    'age' : 'Age',
    'distance' : 'Distance',
    'teachers':'Teachers',
    'poor_prop': 'Poor_Population',
    'airport':'Airport',
    'n_hos_beds' : 'Number_Of_Hospital_Beds',
    'waterbody' : 'Waterbody',
    'rainfall':'Rainfall',  
   'bus_ter':'Bus_Terminal',
    'parks':'Parks',
    'sold':'Sold'}
D1

{'price': 'Price',
 'resid_area': 'Residential_Area',
 'air_qual': 'Air_Qulity',
 'room_num': 'Room_Numbers',
 'age': 'Age',
 'distance': 'Distance',
 'teachers': 'Teachers',
 'poor_prop': 'Poor_Population',
 'airport': 'Airport',
 'n_hos_beds': 'Number_Of_Hospital_Beds',
 'waterbody': 'Waterbody',
 'rainfall': 'Rainfall',
 'bus_ter': 'Bus_Terminal',
 'parks': 'Parks',
 'sold': 'Sold'}

In [9]:
df_houseprice = df_houseprice.rename(columns = D1)
df_houseprice

Price  Residential_Area  Air_Qulity  Room_Numbers   Age  Distance  \
0     24.0             32.31       0.538         6.575  65.2      4.35   
1     21.6             37.07       0.469         6.421  78.9      4.99   
2     34.7             37.07       0.469         7.185  61.1      5.03   
3     33.4             32.18       0.458         6.998  45.8      6.21   
4     36.2             32.18       0.458         7.147  54.2      6.16   
..     ...               ...         ...           ...   ...       ...   
501   22.4             41.93       0.573         6.593  69.1      2.64   
502   20.6             41.93       0.573         6.120  76.7      2.44   
503   23.9             41.93       0.573         6.976  91.0      2.34   
504   22.0             41.93       0.573         6.794  89.3      2.54   
505   19.0             41.93       0.573         6.030  80.8      2.72   

     Teachers  Poor_Population Airport  Number_Of_Hospital_Beds  n_hot_rooms  \
0        24.7             4.98     YES                    5.480      11.1920   
1        22.2             9.14      NO                    7.332      12.1728   
2        22.2             4.03      NO                    7.394     101.1200   
3        21.3             2.94     YES                    9.268      11.2672   
4        21.3             5.33      NO                    8.824      11.2896   
..        ...              ...     ...                      ...          ...   
501      19.0             9.67      NO                    9.348      12.1792   
502      19.0             9.08     YES                    6.612      13.1648   
503      19.0             5.64      NO                    5.478      12.1912   
504      19.0             6.48     YES                    7.940      15.1760   
505      19.0             7.88     YES                   10.280      10.1520   

          Waterbody  Rainfall Bus_Terminal     Parks  Sold  
0             River        23          YES  0.049347     0  
1              Lake        42          YES  0.046146     1  
2               NaN        38          YES  0.045764     0  
3              Lake        45          YES  0.047151     0  
4              Lake        55          YES  0.039474     0  
..              ...       ...          ...       ...   ...  
501  Lake and River        27          YES  0.056006     1  
502  Lake and River        20          YES  0.059903     1  
503             NaN        31          YES  0.057572     1  
504             NaN        47          YES  0.060694     1  
505             NaN        45          YES  0.060336     1  

[506 rows x 16 columns]

In [10]:
df_houseprice.columns

Index(['Price', 'Residential_Area', 'Air_Qulity', 'Room_Numbers', 'Age',
       'Distance', 'Teachers', 'Poor_Population', 'Airport',
       'Number_Of_Hospital_Beds', 'n_hot_rooms', 'Waterbody', 'Rainfall',
       'Bus_Terminal', 'Parks', 'Sold'],
      dtype='object')

### b.null values handling

Laws that we need to fallow while processing the null value 
 - when we have less than 15% data is missing -> remove the observation 
 - between 20% to 70% data is missing -> impure the null value
 - greater than 75% data is missing -> remove the column 

In [11]:
# find the null values 
df_houseprice.isnull().sum()

Price                        0
Residential_Area             0
Air_Qulity                   0
Room_Numbers                 0
Age                          0
Distance                     0
Teachers                     0
Poor_Population              0
Airport                      0
Number_Of_Hospital_Beds      8
n_hot_rooms                  0
Waterbody                  155
Rainfall                     0
Bus_Terminal                 0
Parks                        0
Sold                         0
dtype: int64

In [12]:
#persentage the missing values
round((df_houseprice.isnull().sum())/len(df_houseprice)*100,3)

Price                       0.000
Residential_Area            0.000
Air_Qulity                  0.000
Room_Numbers                0.000
Age                         0.000
Distance                    0.000
Teachers                    0.000
Poor_Population             0.000
Airport                     0.000
Number_Of_Hospital_Beds     1.581
n_hot_rooms                 0.000
Waterbody                  30.632
Rainfall                    0.000
Bus_Terminal                0.000
Parks                       0.000
Sold                        0.000
dtype: float64

In [13]:
df_houseprice.Waterbody.mode()

# this is a categorical a colums so we can replace the null with mode
df_houseprice.Waterbody.fillna('river')


0               River
1                Lake
2               river
3                Lake
4                Lake
            ...      
501    Lake and River
502    Lake and River
503             river
504             river
505             river
Name: Waterbody, Length: 506, dtype: object

In [14]:
df_houseprice.Waterbody.mode()

0    River
Name: Waterbody, dtype: object

In [15]:
# drop a null value
df_houseprice.dropna(inplace =True)


## Prepare data for machine learning

In [16]:
# seperate the target variable and independent variable
target = df_houseprice["Sold"]

In [17]:
#independent variable
df_independent = df_houseprice.drop("Sold", axis=1)

## Seperate  thr categorical colums from independent variables

In [18]:
df_cat = df_independent.select_dtypes(include = 'object')

In [19]:
encoded = pd.get_dummies(df_cat)
encoded


Airport_NO  Airport_YES  Waterbody_Lake  Waterbody_Lake and River  \
0         False         True           False                     False   
1          True        False            True                     False   
3         False         True            True                     False   
4          True        False            True                     False   
6         False         True           False                     False   
..          ...          ...             ...                       ...   
498       False         True           False                     False   
499        True        False            True                     False   
500        True        False            True                     False   
501        True        False           False                      True   
502       False         True           False                      True   

     Waterbody_River  Bus_Terminal_YES  
0               True              True  
1              False              True  
3              False              True  
4              False              True  
6               True              True  
..               ...               ...  
498             True              True  
499            False              True  
500            False              True  
501            False              True  
502            False              True  

[343 rows x 6 columns]

## Scale the numerical data

In [20]:
df_num = df_independent.select_dtypes(include = "number")
df_cat.head()

Airport Waterbody Bus_Terminal
0     YES     River          YES
1      NO      Lake          YES
3     YES      Lake          YES
4      NO      Lake          YES
6     YES     River          YES

In [21]:
#encoded the categorical variable
encoded = pd.get_dummies(df_cat)
encoded

Airport_NO  Airport_YES  Waterbody_Lake  Waterbody_Lake and River  \
0         False         True           False                     False   
1          True        False            True                     False   
3         False         True            True                     False   
4          True        False            True                     False   
6         False         True           False                     False   
..          ...          ...             ...                       ...   
498       False         True           False                     False   
499        True        False            True                     False   
500        True        False            True                     False   
501        True        False           False                      True   
502       False         True           False                      True   

     Waterbody_River  Bus_Terminal_YES  
0               True              True  
1              False              True  
3              False              True  
4              False              True  
6               True              True  
..               ...               ...  
498             True              True  
499            False              True  
500            False              True  
501            False              True  
502            False              True  

[343 rows x 6 columns]

In [22]:
from sklearn.preprocessing import StandardScaler
# initiate the standard scaler
ss = StandardScaler()

In [23]:
# transform the data
Scale = ss.fit_transform(df_num)
Scale

array([[ 0.09978245, -1.22113906, -0.08708422, ..., -0.42960141,
        -1.33138234, -0.42318421],
       [-0.16181282, -0.52970502, -0.68208471, ..., -0.18690803,
         0.19056446, -0.72689187],
       [ 1.12436392, -1.24002276, -0.77693986, ..., -0.4109936 ,
         0.43087185, -0.63156184],
       ...,
       [-0.68500335, -0.14912577,  0.31820597, ...,  0.29847872,
         1.15179402, -0.12437127],
       [-0.07461439,  0.17625495,  0.21472763, ..., -0.18532439,
        -1.01097249,  0.20842292],
       [-0.27081085,  0.17625495,  0.21472763, ...,  0.05855672,
        -1.57168973,  0.57815884]])

In [24]:
df_Scale = pd.DataFrame(Scale,columns = df_num.columns)
df_Scale = df_Scale.reset_index()
df_Scale

index     Price  Residential_Area  Air_Qulity  Room_Numbers       Age  \
0        0  0.099782         -1.221139   -0.087084      0.351278 -0.049985   
1        1 -0.161813         -0.529705   -0.682085      0.144198  0.425106   
2        2  1.124364         -1.240023   -0.776940      0.920077 -0.722742   
3        3  1.429558         -1.240023   -0.776940      1.120434 -0.431445   
4        4 -0.020115         -0.413498   -0.207809     -0.405775 -0.001436   
..     ...       ...               ...         ...           ...       ...   
338    338 -0.205412         -0.149126    0.318206     -0.396362 -0.046517   
339    339 -0.608705         -0.149126    0.318206     -1.001467  0.237844   
340    340 -0.685003         -0.149126    0.318206     -0.385605  0.452849   
341    341 -0.074614          0.176255    0.214728      0.375483  0.085260   
342    342 -0.270811          0.176255    0.214728     -0.260550  0.348814   

     Distance  Teachers  Poor_Population  Number_Of_Hospital_Beds  \
0    0.139406  1.384755        -1.015715                -1.606757   
1    0.441920  0.233366        -0.423275                -0.378839   
2    1.018587 -0.181134        -1.306238                 0.904773   
3    0.994953 -0.181134        -0.965870                 0.610391   
4    0.857877  1.430811         0.045265                -0.626810   
..        ...       ...              ...                      ...   
338 -0.649967 -0.411412         0.115048                -1.378677   
339 -0.753956 -0.411412         0.425509                 1.290652   
340 -0.668874 -0.411412         0.315851                -1.105512   
341 -0.668874 -1.240412        -0.347796                 0.957814   
342 -0.763410 -1.240412        -0.431820                -0.856215   

     n_hot_rooms  Rainfall     Parks  
0      -0.429601 -1.331382 -0.423184  
1      -0.186908  0.190564 -0.726892  
2      -0.410994  0.430872 -0.631562  
3      -0.405451  1.231896 -1.359756  
4      -0.184335  0.110462 -0.155435  
..           ...       ...       ...  
338     0.307189 -0.610460  0.069944  
339     0.299864  0.591077  0.916130  
340     0.298479  1.151794 -0.124371  
341    -0.185324 -1.010972  0.208423  
342     0.058557 -1.571690  0.578159  

[343 rows x 13 columns]

In [25]:
X = pd.concat([df_num,encoded],axis = 1)
X

Price  Residential_Area  Air_Qulity  Room_Numbers   Age  Distance  \
0     24.0             32.31       0.538         6.575  65.2      4.35   
1     21.6             37.07       0.469         6.421  78.9      4.99   
3     33.4             32.18       0.458         6.998  45.8      6.21   
4     36.2             32.18       0.458         7.147  54.2      6.16   
6     22.9             37.87       0.524         6.012  66.6      5.87   
..     ...               ...         ...           ...   ...       ...   
498   21.2             39.69       0.585         6.019  65.3      2.68   
499   17.5             39.69       0.585         5.569  73.5      2.46   
500   16.8             39.69       0.585         6.027  79.7      2.64   
501   22.4             41.93       0.573         6.593  69.1      2.64   
502   20.6             41.93       0.573         6.120  76.7      2.44   

     Teachers  Poor_Population  Number_Of_Hospital_Beds  n_hot_rooms  \
0        24.7             4.98                    5.480      11.1920   
1        22.2             9.14                    7.332      12.1728   
3        21.3             2.94                    9.268      11.2672   
4        21.3             5.33                    8.824      11.2896   
6        24.8            12.43                    6.958      12.1832   
..        ...              ...                      ...          ...   
498      20.8            12.92                    5.824      14.1696   
499      20.8            15.10                    9.850      14.1400   
500      20.8            14.33                    6.236      14.1344   
501      19.0             9.67                    9.348      12.1792   
502      19.0             9.08                    6.612      13.1648   

     Rainfall     Parks  Airport_NO  Airport_YES  Waterbody_Lake  \
0          23  0.049347       False         True           False   
1          42  0.046146        True        False            True   
3          45  0.047151       False         True            True   
4          55  0.039474        True        False            True   
6          41  0.052170       False         True           False   
..        ...       ...         ...          ...             ...   
498        32  0.054546       False         True           False   
499        47  0.063466        True        False            True   
500        54  0.052497        True        False            True   
501        27  0.056006        True        False           False   
502        20  0.059903       False         True           False   

     Waterbody_Lake and River  Waterbody_River  Bus_Terminal_YES  
0                       False             True              True  
1                       False            False              True  
3                       False            False              True  
4                       False            False              True  
6                       False             True              True  
..                        ...              ...               ...  
498                     False             True              True  
499                     False            False              True  
500                     False            False              True  
501                      True            False              True  
502                      True            False              True  

[343 rows x 18 columns]

In [26]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X, target, random_state = 10 , test_size = 0.2)
X_train.shape
y_train.shape

X_test.shape
X_test.shape

(69, 18)

In [34]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

tree_model = DecisionTreeClassifier()

tree_model.fit(X_train,y_train)

prediction = tree_model.predict(X_test)

accuracy = accuracy_score(y_test, prediction)
print("accuracy:", accuracy)

accuracy: 0.6811594202898551
